### ISSUE TITLE
PCRs method doesn't allow providing hash algorithm

### ISSUE THREAD

**Issue by ericchiang** [google/go-attestation#72]

**Comment by ericchiang** [comment]
Just because PCRs can provide SHA256, doesn't mean that all events contain SHA256 values. It's also unclear if the quote signature is tied to this hash value.

We need to figure out a better strategy for SHA1 vs SHA256.

**Comment by ericchiang** [comment]
We hit an issue with shielded VMs where the PCRs provided SHA256 values but weren't logged in the event log. The event log is only providing SHA1 digest. It's unclear if this is a bug in GCP or our code.

**Comment by ericchiang** [comment]
For the test case in #64 from GCP

Event log

```
0       1489f923c4dca729178b3e3233458550d8dddf29
7       d4fdd1f14d4041494deb8fc990c45343d2277d08
7       5abd9412abf33e34a79b3d1a93d350e742d8ecd8
7       f0501c79b607cc42e9142ee85a74d9c27669c0e2
7       a0e46611f6906ab3c0674d8971b0e4d9ea504ce4
7       9e04b683b1ade74270dc6083dd716acc63a33310
7       9069ca78e7450a285173431b3e52c5c25299e473
7       b893de4a83f078b42dc089b4bd6cc7aa5b128c05
5       6c1ecadf12a19582e80d66c7773f521c4193afe9
4       57a3e40bae6ae5ab1427c6aff22aa4f06e158ef4
11      5497b0911b3f5772723def3b360a2e654327c19b
12      74b8480c3c82b3e76ff72a09db378230c67388fd
13      ca2bc43b9555a851bf767876493668f892ef7319
14      01fd60a7193434b25ee8870827fd436b125aa03d
12      f45b936292f6f64ad639819a1368052486bfc7d1
13      d8f11c636a61f54d3c3cce9b8e7da89f14033c02
14      e4ea7b40b3bf9b57183b5e85e58459fb76e449b0
11      3a4072cc6b77e2639d4fdc91c91efc11bc3e33c3
12      9d7f499388daa8e7d7f1e399616e39e5891d399d
13      9d7f499388daa8e7d7f1e399616e39e5891d399d
14      9d7f499388daa8e7d7f1e399616e39e5891d399d
```

PCRs

```
0       51c323de0c0c694f4601cdd02beb58ff13629f74
1       0000000000000000000000000000000000000000
2       0000000000000000000000000000000000000000
3       0000000000000000000000000000000000000000
4       0ca4b4a4784bf4eed9c3556aba1dac5585a5951a
5       2b022297d4f1e0101c8c986be229c8dd0350514d
6       0000000000000000000000000000000000000000
7       859a5877266b5c909613468091a73380a5386786
8       0000000000000000000000000000000000000000
9       0000000000000000000000000000000000000000
10      0000000000000000000000000000000000000000
11      ebb98df76613280f20dc38221143a9e727399486
12      75f3e16b6ef0b455282ed8fbbdfcc3da9abd241d
13      383de79fbdde6296205e2afe44800e0c053fc82f
14      275a689f9d5f8244a4b999fabe600c5816be5511
15      0000000000000000000000000000000000000000
16      0000000000000000000000000000000000000000
17      ffffffffffffffffffffffffffffffffffffffff
18      ffffffffffffffffffffffffffffffffffffffff
19      ffffffffffffffffffffffffffffffffffffffff
20      ffffffffffffffffffffffffffffffffffffffff
21      ffffffffffffffffffffffffffffffffffffffff
22      ffffffffffffffffffffffffffffffffffffffff
23      0000000000000000000000000000000000000000
```

**Comment by josephlr** [comment]
> We hit an issue with shielded VMs where the PCRs provided SHA256 values but weren't logged in the event log. The event log is only providing SHA1 digest. It's unclear if this is a bug in GCP or our code.

Ya, this looks like a GCP related issue. Will look into it tomorrow

**Comment by ericchiang** [comment]
@josephlr fyi, the event log replay was our issue. It correctly replays after changes in #64 :)

It'd still be interesting to know if the event log only containing SHA1 was our issues or GCP's. We'll try to look into it more tomorrow too.

**Comment by josephlr** [comment]
> @josephlr fyi, the event log replay was our issue. It correctly replays after changes in #64 :)
> 
> It'd still be interesting to know if the event log only containing SHA1 was our issues or GCP's. We'll try to look into it more tomorrow too.

Sorry, meant to say that the GCP firmware right now only logs events w/ SHA1 and not SHA256. We should probably have it do both. We want customers to be entirely SHA1-free, if they so desire.

**Comment by ericchiang** [comment]
cc @mjg59 ^

**Comment by ericchiang** [comment]
> I think this might just be a limitation of the Linux Kernel though, as it can't process the crypto-agile format that contains those logs.

@josephlr this is a Windows VM :)

**Comment by ericchiang** [comment]
Closed by #77

The Shielded VM issue still isn't resolved, but going to open something internally against GCP.
